Game/Source/GDKTestGyms/BenchmarkGymGameMode.cpp

```diff
@@ -145,18 +175,72 @@ void ABenchmarkGymGameMode::Tick(float DeltaSeconds)
 
 		for (int i = AIControlledPlayers.Num() - 1; i >= 0; i--)
 		{
-			checkf(AIControlledPlayers[i].Controller, TEXT("Simplayer controller has been deleted."));
-			ACharacter* Character = AIControlledPlayers[i].Controller->GetCharacter();
+			AController* Controller = AIControlledPlayers[i].Key.Get();
+			int InfoIndex = AIControlledPlayers[i].Value;
+
+			checkf(Controller, TEXT("Simplayer controller has been deleted."));
+			ACharacter* Character = Controller->GetCharacter();
 			checkf(Character, TEXT("Simplayer character does not exist."));
-			int InfoIndex = AIControlledPlayers[i].Index;
 			UDeterministicBlackboardValues* Blackboard = Cast<UDeterministicBlackboardValues>(Character->FindComponentByClass(UDeterministicBlackboardValues::StaticClass()));
 			checkf(Blackboard, TEXT("Simplayer does not have a UDeterministicBlackboardValues component."));
-			
+
 			const FBlackboardValues& Points = PlayerRunPoints[InfoIndex % PlayerRunPoints.Num()];
 			Blackboard->ClientSetBlackboardAILocations(Points);
 		}
 		AIControlledPlayers.Empty();
 	}
+
+	ServerUpdateNFRTestMetrics(DeltaSeconds);
+}
+
+void ABenchmarkGymGameMode::ServerUpdateNFRTestMetrics(float DeltaSeconds)
+{
+	if (GetNumPlayers() != ExpectedPlayers) // Wait for a complete test scenario
+	{
+		return;
+	}
+
+	float ClientRTTSeconds = 0.0f;
+	int ClientRTTCount = 0;
+	float ClientViewLatenessSeconds = 0.0f;
+	int ClientViewLatenessNum = 0;
+	for (TObjectIterator<UUserExperienceReporter> Itr; Itr; ++Itr) // These exist on player characters
+	{
+		UUserExperienceReporter* Component = *Itr;
+		if(Component->GetOwner() != nullptr && Component->GetWorld() == GetWorld())
+		{
+			ClientRTTSeconds += Component->ServerRTT;
+			ClientRTTCount++;
+
+			ClientViewLatenessSeconds += Component->ServerViewLateness;
+			ClientViewLatenessNum++;
+		}
+	}
+	ClientRTTSeconds /= static_cast<float>(ClientRTTCount) + 0.00001f; // Avoid div 0
+	ClientViewLatenessSeconds /= static_cast<float>(ClientViewLatenessNum) + 0.00001f; // Avoid div 0
+
+	AveragedClientRTTSeconds = ClientRTTSeconds;
```

Comment:
If ClientRTTCount or ClientViewLatenessNum are less than ExpectedPlayers something's gone wrong - might be worth logging

Reply:
How should we handle that, just force a failure of the metric?

Reply:
Just add a warning for now - we can inspect the logs and see if it comes up